File: app/src/main/java/com/kieran/winnipegbus/Activities/TripPlannerActivity.java
package com.kieran.winnipegbus.Activities;

import android.app.DatePickerDialog;
import android.app.TimePickerDialog;
import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.View;
import android.widget.Adapter;
import android.widget.AdapterView;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.ExpandableListView;
import android.widget.Spinner;
import android.widget.TimePicker;

import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.location.LocationServices;
import com.kieran.winnipegbus.Adapters.TripPlannerAdapter;
import com.kieran.winnipegbus.LocationPickerDialog;
import com.kieran.winnipegbus.R;
import com.kieran.winnipegbusbackend.LoadResult;
import com.kieran.winnipegbusbackend.StopTime;
import com.kieran.winnipegbusbackend.TransitApiManager;
import com.kieran.winnipegbusbackend.TripPlanner.TimeMode;
import com.kieran.winnipegbusbackend.TripPlanner.classes.Location;
import com.kieran.winnipegbusbackend.TripPlanner.classes.Trip;
import com.kieran.winnipegbusbackend.TripPlanner.classes.TripParameters;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.GregorianCalendar;
import java.util.List;

public class TripPlannerActivity extends GoogleApiActivity implements TransitApiManager.OnJsonLoadResultReceiveListener {
    private static final String PARAMETERS = "parameters";
    private TripParameters tripParameters = new TripParameters();
    private List<Trip> trips;
    private TripPlannerAdapter adapter;
    private Button getDirectionsButton;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_trip_planner);
        Intent intent = getIntent();

        tripParameters = (TripParameters) intent.getSerializableExtra(PARAMETERS);

        if(tripParameters == null)
            tripParameters = new TripParameters();

        trips = new ArrayList<>();
        ExpandableListView listView = (ExpandableListView) findViewById(R.id.trip_planner_listview);
        listView.addHeaderView(getLayoutInflater().inflate(R.layout.listview_trip_planner_header, null));
        adapter = new TripPlannerAdapter(this, trips);
        listView.setAdapter(adapter);

        getDirectionsButton = (Button) findViewById(R.id.get_directions_button);

        initializeFields();

        googleApiClient = new GoogleApiClient.Builder(this)
                .addApi(LocationServices.API)
                .addConnectionCallbacks(this)
                .addOnConnectionFailedListener(this).build();
        connectClient();
    }

    private void initializeFields() {
        if (tripParameters.getOrigin() != null)
            ((Button)findViewById(R.id.origin_select_button)).setText(tripParameters.getOrigin().getTitle());
        if (tripParameters.getDestination() != null)
            ((Button)findViewById(R.id.destination_select_button)).setText(tripParameters.getDestination().getTitle());

        setTextViewText(R.id.trip_time_hour_minute, tripParameters.getTime().toFormattedString(null, getTimeSetting()));
        setTextViewText(R.id.trip_time_date, tripParameters.getTime().toDatePickerDateFormat());

        Spinner timeModeView = (Spinner) findViewById(R.id.time_mode_spinner);
        final Adapter adapter = timeModeView.getAdapter();

        for(int i = 0; i < adapter.getCount(); i++) {
            if(((String)adapter.getItem(i)).equals(tripParameters.getTimeMode().name)) {
                timeModeView.setSelection(i);
            }
        }

        timeModeView.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                tripParameters.setTimeMode(TimeMode.getTimeModeByName((String) adapter.getItem(position)));
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {

            }
        });

        getDirectionsButton.setEnabled(tripParameters.isValid());
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_trip_planner, menu);
        return true;
    }

    public void getDirections(View view) {
        trips.clear();
        TransitApiManager.getJsonAsync(tripParameters.getURL(), this);
    }

    public void pickTime(View view) {
        final Calendar c = new GregorianCalendar();
        TimePickerDialog d = new TimePickerDialog(this, new TimePickerDialog.OnTimeSetListener() {
            @Override
            public void onTimeSet(TimePicker view, int hourOfDay, int minute) {
                c.setTimeInMillis(tripParameters.getTime().getMilliseconds());
                c.set(Calendar.HOUR_OF_DAY, hourOfDay);
                c.set(Calendar.MINUTE, minute);
                tripParameters.setTime(new StopTime(c.getTimeInMillis()));

                setTextViewText(R.id.trip_time_hour_minute, tripParameters.getTime().toFormattedString(null, getTimeSetting()));
            }
        }, tripParameters.getTime().getHours(), tripParameters.getTime().getMinutes(), getTimeSetting());

        d.setTitle("");
        d.show();
    }

    public void pickDate(View view) {
        final Calendar c = new GregorianCalendar();
        final DatePickerDialog d = new DatePickerDialog(this, new DatePickerDialog.OnDateSetListener() {
            @Override
            public void onDateSet(DatePicker view, int year, int monthOfYear, int dayOfMonth) {
                c.setTimeInMillis(tripParameters.getTime().getMilliseconds());
                c.set(Calendar.YEAR, year);
                c.set(Calendar.MONTH, monthOfYear);
                c.set(Calendar.DAY_OF_MONTH, dayOfMonth);
                tripParameters.setTime(new StopTime(c.getTimeInMillis()));

                setTextViewText(R.id.trip_time_date, tripParameters.getTime().toDatePickerDateFormat());
            }
        }, tripParameters.getTime().getYear(), tripParameters.getTime().getMonth(), tripParameters.getTime().getDayOfMonth());

        d.setTitle("");
        d.show();
    }

    @Override
    public void OnReceive(LoadResult<JSONObject> result) {
        JSONObject jsonObject = result.getResult();
        try {
            JSONArray plans = jsonObject.getJSONArray("plans");



            for(int i = 0; i < plans.length(); i++) {
                trips.add(new Trip(plans.getJSONObject(i)));
            }

            adapter.notifyDataSetChanged();
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void onConnected(Bundle bundle) {
        requestLocation();
    }

    public void selectOrigin(View view) {
        new LocationPickerDialog(this, new LocationPickerDialog.OnLocationPickedListener() {
            @Override
            public void OnLocationPickedListener(Location location) {
                tripParameters.setOrigin(location);
                initializeFields();
            }
        }).show();
    }

    public void selectDestination(View view) {
        new LocationPickerDialog(this, new LocationPickerDialog.OnLocationPickedListener() {
            @Override
            public void OnLocationPickedListener(Location location) {
                tripParameters.setDestination(location);
                initializeFields();
            }
        }).show();
    }

    public void swapLocations(View view) {
        tripParameters.swapLocations();
        initializeFields();
    }
}

File: app/src/main/java/com/kieran/winnipegbus/Activities/TripPlannerActivity.kt
package com.kieran.winnipegbus.Activities

import android.app.DatePickerDialog
import android.app.TimePickerDialog
import android.content.Intent
import android.os.Bundle
import android.support.v4.widget.SwipeRefreshLayout
import android.view.Menu
import android.view.View
import android.widget.Adapter
import android.widget.AdapterView
import android.widget.Button
import android.widget.DatePicker
import android.widget.ExpandableListView
import android.widget.Spinner
import android.widget.TimePicker

import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.location.LocationServices
import com.kieran.winnipegbus.Adapters.TripPlannerAdapter
import com.kieran.winnipegbus.LocationPickerDialog
import com.kieran.winnipegbus.R
import com.kieran.winnipegbus.Views.StyledSwipeRefresh
import com.kieran.winnipegbusbackend.LoadResult
import com.kieran.winnipegbusbackend.StopTime
import com.kieran.winnipegbusbackend.TransitApiManager
import com.kieran.winnipegbusbackend.TripPlanner.TimeMode
import com.kieran.winnipegbusbackend.TripPlanner.classes.Location
import com.kieran.winnipegbusbackend.TripPlanner.classes.Trip
import com.kieran.winnipegbusbackend.TripPlanner.classes.TripParameters

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

import java.util.ArrayList
import java.util.Calendar
import java.util.GregorianCalendar

class TripPlannerActivity : GoogleApiActivity(), TransitApiManager.OnJsonLoadResultReceiveListener, SwipeRefreshLayout.OnRefreshListener {
    private var tripParameters: TripParameters? = TripParameters()
    private var trips: MutableList<Trip>? = null
    private var adapter: TripPlannerAdapter? = null
    private var getDirectionsButton: Button? = null
    private var swipeRefresh: StyledSwipeRefresh? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_trip_planner)
        val intent = intent

        tripParameters = intent.getSerializableExtra(PARAMETERS) as TripParameters

        if (tripParameters == null)
            tripParameters = TripParameters()

        trips = ArrayList()
        val listView = findViewById<View>(R.id.trip_planner_listview) as ExpandableListView
        listView.addHeaderView(layoutInflater.inflate(R.layout.listview_trip_planner_header, null))
        adapter = TripPlannerAdapter(this, trips)
        listView.setAdapter(adapter)

        getDirectionsButton = findViewById<View>(R.id.get_directions_button) as Button
        swipeRefresh = findViewById<View>(R.id.trip_planner_swipe_refresh) as StyledSwipeRefresh
        swipeRefresh!!.setOnRefreshListener(this)
        initializeFields()

        googleApiClient = GoogleApiClient.Builder(this)
                .addApi(LocationServices.API)
                .addConnectionCallbacks(this)
                .addOnConnectionFailedListener(this).build()
        connectClient()
    }

    private fun initializeFields() {
        val originButton = findViewById<View>(R.id.origin_select_button) as Button
        val destinationButton = findViewById<View>(R.id.destination_select_button) as Button

        if (tripParameters!!.origin != null)
            originButton.text = tripParameters!!.origin!!.title
        else
            originButton.text = "Select"

        if (tripParameters!!.destination != null)
            destinationButton.text = tripParameters!!.destination!!.title
        else
            destinationButton.text = "Select"

        setTextViewText(R.id.trip_time_hour_minute, tripParameters!!.time!!.toFormattedString(null, timeSetting))
        setTextViewText(R.id.trip_time_date, tripParameters!!.time!!.toDatePickerDateFormat())

        val timeModeView = findViewById<View>(R.id.time_mode_spinner) as Spinner
        val adapter = timeModeView.adapter

        for (i in 0 until adapter.count) {
            if (adapter.getItem(i) as String == tripParameters!!.timeMode!!.name) {
                timeModeView.setSelection(i)
            }
        }

        timeModeView.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
                tripParameters!!.timeMode = TimeMode.getTimeModeByName(adapter.getItem(position) as String)
            }

            override fun onNothingSelected(parent: AdapterView<*>) {

            }
        }

        getDirectionsButton!!.isEnabled = tripParameters!!.isValid
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_trip_planner, menu)
        return true
    }

    fun getDirections(view: View) {
        swipeRefresh!!.isRefreshing = true
        getDirections()
    }

    fun getDirections() {
        trips!!.clear()
        TransitApiManager.getJsonAsync(tripParameters!!.url, this)
    }

    fun pickTime(view: View) {
        val c = GregorianCalendar()
        val d = TimePickerDialog(this, TimePickerDialog.OnTimeSetListener { view, hourOfDay, minute ->
            c.timeInMillis = tripParameters!!.time!!.milliseconds
            c.set(Calendar.HOUR_OF_DAY, hourOfDay)
            c.set(Calendar.MINUTE, minute)
            tripParameters!!.time = StopTime(c.timeInMillis)

            setTextViewText(R.id.trip_time_hour_minute, tripParameters!!.time!!.toFormattedString(null, timeSetting))
        }, tripParameters!!.time!!.hours, tripParameters!!.time!!.minutes, timeSetting)

        d.setTitle("")
        d.show()
    }

    fun pickDate(view: View) {
        val c = GregorianCalendar()
        val d = DatePickerDialog(this, DatePickerDialog.OnDateSetListener { view, year, monthOfYear, dayOfMonth ->
            c.timeInMillis = tripParameters!!.time!!.milliseconds
            c.set(Calendar.YEAR, year)
            c.set(Calendar.MONTH, monthOfYear)
            c.set(Calendar.DAY_OF_MONTH, dayOfMonth)
            tripParameters!!.time = StopTime(c.timeInMillis)

            setTextViewText(R.id.trip_time_date, tripParameters!!.time!!.toDatePickerDateFormat())
        }, tripParameters!!.time!!.year, tripParameters!!.time!!.month, tripParameters!!.time!!.dayOfMonth)

        d.setTitle("")
        d.show()
    }

    override fun OnReceive(result: LoadResult<JSONObject>) {
        val jsonObject = result.result
        try {
            val plans = jsonObject!!.getJSONArray("plans")

            for (i in 0 until plans.length()) {
                trips!!.add(Trip(plans.getJSONObject(i)))
            }

            adapter!!.notifyDataSetChanged()
            swipeRefresh!!.isRefreshing = false
        } catch (e: JSONException) {

        }

    }

    override fun onConnected(bundle: Bundle) {
        requestLocation()
    }

    fun selectOrigin(view: View) {
        LocationPickerDialog(this, LocationPickerDialog.OnLocationPickedListener { location ->
            tripParameters!!.origin = location
            initializeFields()
        }).show()
    }

    fun selectDestination(view: View) {
        LocationPickerDialog(this, LocationPickerDialog.OnLocationPickedListener { location ->
            tripParameters!!.destination = location
            initializeFields()
        }).show()
    }

    fun swapLocations(view: View) {
        tripParameters!!.swapLocations()
        initializeFields()
    }

    override fun onRefresh() {
        getDirections()
    }

    companion object {
        private val PARAMETERS = "parameters"
    }
}